### PR TITLE
amp based zeroing for the winch and arm extension and arm rotation with timout

### PR DESCRIPTION
**Pull request by MidgleyM**

### CODE REVIEW

File: robotbase/src/main/java/com/team2357/frc2023/commands/armextension/ArmExtendAmpZeroCommand.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    
+    @Override
+    public boolean isFinished() {
+        return ArmExtensionSubsystem.getInstance().getAmps() >= Constants.ARM_EXTENSION.ARM_EXTENSION_AMP_ZERO_MAX_AMPS;
+    }
+    
+    @Override
+    public void end(boolean interrupted) {
+        ArmExtensionSubsystem.getInstance().stopMotor();
+        ArmExtensionSubsystem.getInstance().resetEncoder();
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2023/commands/armextension/ArmExtendAmpZeroCommand.java:26]: I think you probably don't want to reset the encoder if `interrupted == true`

File: robotbase/src/main/java/com/team2357/frc2023/commands/armrotation/ArmRotationAmpZeroCommand.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+@Override
+public boolean isFinished() {
+    return ArmRotationSubsystem.getInstance().getAmps() >= Constants.ARM_ROTATION.ARM_ROTATION_AMP_ZERO_MAX_AMPS;
+}
+
+@Override
+public void end(boolean interrupted) {
+    ArmRotationSubsystem.getInstance().stopRotationMotors();
+    ArmRotationSubsystem.getInstance().resetEncoders();
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2023/commands/armrotation/ArmRotationAmpZeroCommand.java:26]: Same comment on reset while `interrupted == true`

File: robotbase/src/main/java/com/team2357/frc2023/commands/intake/WinchAmpZeroCommand.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    
+    @Override
+    public boolean isFinished() {
+        return IntakeArmSubsystem.getInstance().getAmps() >= Constants.INTAKE_ARM.WINCH_AMP_ZERO_MAX_AMPS;
+    }
+    
+    @Override
+    public void end(boolean interrupted) {
+        IntakeArmSubsystem.getInstance().stopWinchMotor();
+        IntakeArmSubsystem.getInstance().resetEncoders();
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2023/commands/intake/WinchAmpZeroCommand.java:27]: Same comment on reset

File: robotbase/src/main/java/com/team2357/frc2023/commands/util/AmpZeroCommandGroup.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import com.team2357.frc2023.commands.armrotation.ArmRotationAmpZeroCommand;
+import com.team2357.frc2023.commands.intake.WinchAmpZeroCommand;
+import com.team2357.frc2023.commands.armextension.ArmExtendAmpZeroCommand;
+
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.ParallelDeadlineGroup;
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+
+public class AmpZeroCommandGroup extends ParallelDeadlineGroup{
+    static Command whinchZero = new WinchAmpZeroCommand();
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2023/commands/util/AmpZeroCommandGroup.java:13]: Misspelling, `whinch` should be `winch`

File: robotbase/src/main/java/com/team2357/frc2023/commands/util/AmpZeroCommandGroup.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    static Command armRotationZero = new ArmRotationAmpZeroCommand();
+
+    public AmpZeroCommandGroup() {
+        super(new WaitCommand(Constants.AMP_ZERO.AMP_ZERO_DEADLINE_SECONDS), new WinchAmpZeroCommand(), new ArmRotationAmpZeroCommand(), new ArmExtendAmpZeroCommand());
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2023/commands/util/AmpZeroCommandGroup.java:17]: Keep in mind if the wait is exceeded, one or more of these commands will fail to completely zero. I think we should at the least log an error that the zero command didn't complete for each one that doesn't.

**Review by coderkevin — COMMENTED** [review]
One main question on what happens if the wait is exceeded and the zero command(s) are interrupted

**Review by NolanCampbell2357 — APPROVED** [review]

**Pull request merged**